### DISCUSSION TITLE
How Are Platform-Specific Extensions Handled When Run In Unsupported Platforms?

### DISCUSSION THREAD

**Discussion by ericdrobinson** [microsoft/vscode-discussions#19] [Extension Development QnA] [answered]
I am working on an extension that includes a native node module. Because of dependencies for which I do not have source access, the extension is limited to the following platforms:

- `win32-ia32`
- `win32-x64`
- `darwin-x64`

Notably missing is the `darwin-arm64` platform (I expect `win32-arm64` to also become an issue soon).

According to the documentation for [Platform-specific extensions](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions):

> If no package has been published for a platform, the user will see the extension appear as disabled and it can not be installed.

Okay, cool. That's helpful. But... what happens if the user:

1. Starts up VS Code in Rosetta
2. Installs an extension that only works on `darwin-x64`
3. Restarts VS Code in native Apple Silicon mode (non-Rosetta)

or

1. Downloads the extension from the Marketplace directly
2. Manually installs the extension

The documentation makes no mention of what happens in these cases. It seems that there are two likely scenarios:

1. The extension will `activate()` but functionality that requires the native node module will [possibly silently] fail.
2. The extension will be disabled when run in a VS Code instance that is itself running an unsupported architecture.

Does anyone know what such extensions do in these circumstances?

**Comment by isidorn** [comment, marked as answer]
@ericdrobinson thanks for opening this discussion.
Case 1 - we do not support this. So the extension will be downloaded for `darwin-x64`. When you restart the extension will still be `darwin-x64`. We do not detect this.
Case 2 - users can do this and manually install extensions for the wrong platform. This will probably lead to a broken experience since extensions will probably not work on the wrong platform due to native node modules.

So the extension will `activate()` but functionality that requires native node modules will fail.

To the best of my knowledge, extensions can not do much about these cases. These are issues on the vscode side. However they seem to happen rarely since we got 0 issues filed for this by users.

@sandy081 can correct me if I was wrong.

**Reply by ericdrobinson** [reply]
Oh man, thanks so much for the clarifications! Is there any way this info can make its way into [the documentation](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions)? As written, the documentation is not clear that the language refers to publication specifically. Understanding that this workflow does not completely protect extension developers from having to help users with a "bad installation" is very important.

**Reply by ericdrobinson** [reply]
> These are issues on the vscode side. However they seem to happen rarely since we got 0 issues filed for this by users.

@isidorn I doubt you would. It has been our experience that users who run into issues here reach out to _us_ and tell us that our extension is broken or doesn't work. Even though the root of the problem is in VS Code, it is the extension developers who face the feedback.

> To the best of my knowledge, extensions can not do much about these cases.

We have determined that we can at least alert users who attempt to use the extension on an unsupported platform and then "stop" (we think?) the activation of the extension. Specifically, we:

1. Identify the platform the VS Code instance is running on. We currently do this using the following code:
   ```ts
   function GetCurrentPlatform(): string
   {
       return `${process.platform}-${process.arch}`;
   }
   ```
   This code will return platform strings that match the tags outlined in the [Platform-specific extensions](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions) documentation.
2. If a non-supported platform is detected, we:
   1. Show a warning notification to the user:
      ```ts
      const msg: string = "The [NAME] extension does not currently run natively on this device. For more information please see the extension's README.";
      const readmeButtomText: string = "View README";
      vscode.window.showWarningMessage(msg, readmeButtomText).then(/** Handle button **/);
      ```
      _[The "View README" button is co takes the user to the extension's marketplace page (and possibly a specific header therein).]_
   2. Throw an error to signal to VS Code that activation failed:
      ```ts
      throw new Error("The [NAME] extension is not supported on the current architecture.");
      ```

While we've not tested this workflow with users yet, in local tests it does appear to help provide actionable information to users and stops the extension from constantly trying (and failing) to work.

It may be worthwhile suggesting this sort of approach as something extension developers can/should implement in the event that a platform is unsupported. Alternatively, you could implement a system that Extension developers could connect to (e.g. a list of supported platforms in the `package.json` manifest?) and then add consistent messaging from VS Code's side when it detects a mismatch between extension and operating platform...

**Comment by damms005** [comment]
Hi @ericdrobinson

Please can you share with me how you achieved this cross-platform extension publishing?

In [my extension](https://marketplace.visualstudio.com/items?itemName=damms005.devdb), I am using [sqlite3](https://www.npmjs.com/package/sqlite3) which contains native code. When I build and publish the extension from my Linux machine, [bad things happen](https://github.com/damms005/devdb-vscode/issues/46) and now I know why.

To fix this, my guess is that I need to use GitHub actions to build and publish the extension using the [Platform-specific extensions](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions) publishing feature by running `vsce publish --target win32-x64 win32-arm64` as a step in the CI job for Windows for example. I am thinking I will use [matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) for this.

Is this the kind of workflow you adopted or I am missing something?

**Reply by gjsjohnmurray** [reply]
@damms005 there is some information at https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions including links to a sample.

**Reply by damms005** [reply]
Thank you. I appreciate the response :pray:
